tests/unit/test_history_injection.py

```
@@ -0,0 +1,280 @@
+        ku.validate_chat_history([{"role": "assistant", "tool_calls": [{"id": None}]}])
+
+def test_validate_history_tool_call_invalid_type():
+    with pytest.raises(ValueError, match="'type' must be 'function'"):
+        ku.validate_chat_history([{"role": "assistant", "tool_calls": [{"id": "1", "type": "not_fn"}]}])
+
+def test_validate_history_tool_call_function_not_dict():
+    with pytest.raises(ValueError, match="'function' must be a dict"):
+        ku.validate_chat_history([{"role": "assistant", "tool_calls": [{"id": "1", "type": "function", "function": "not a dict"}]}])
+
+def test_validate_history_tool_call_function_name_missing():
+    with pytest.raises(ValueError, match="'function.name' must be a non-empty string"):
```

Ruff (0.15.12)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)